### PR TITLE
DEV: Re-Organise Loading of Bundled Resources & Remote Samples

### PR DESCRIPTION
**Pull request by John-P**
# Summary of Changes

- Moved local test files into `tiatoolbox/data`.
- Fixed `Manifest.ini` and added  `tiatoolbox/data`. This means that this directory will be downloaded with the package.
- Using `pkg_resources` to properly load bundled resources (e.g. `target_image.png`) in `tiatoolbox.data`. Bundled sample files can now be loaded via `tiatoolbox.data` e.g. `img = tiatoolbox.data.stainnorm_target()`. This is much the same as in `skimage`. See `tiatoolbox/data/__init__.py` for more. This is implemented using a function `_local_sample_path` to find the path to files bundled with the module.
- Removed duplicate code in `conftest.py` for downloading remote files. This is now in `tiatoolbox.data._fetch_remote_file` which will load a remote resource using the corresponding key in `tiatoolbox/data`. There is a new factory fixture to load a remote resource by name (string) from the YAML manifest (`tiatoolbox/data/remote_samples.yaml`). This is used to replace the duplicate code in the sample file fixtures. These sample fixtures could potentially be removed and use this factory ditrectly. There is one slight advantage to creating fixteres in this way in that there can be helpful and descriptive docstrings and type annotations (current docstrings are all the same and not very desriptive though). A good example can be seen in `skimage.data`.

# To-Dos

- [x] Fix errors raised by new flake8 rules.
  - [x]  Remove leading underscores from fixtures.
- [x] Fix Deepsource Grumbles
- [x] Check for anything breaking.
- [x] Rename some remote sample files to make more sense (either just the key in the YAML or the actual file too).

# Renaming Sample Files

Some of the remote sample files do not have helpful file names or they are inconsistent. Additionally, the docstrings for the pytext fixtures for loading these files are note descriptive. We should do an audit of this to improve maintainability. 

# Considerations

We could switch to using the [pooch](https://www.fatiando.org/pooch/latest/) package (used in skimage) to download remote resources. What do you think @shaneahmed? Shall I replace our own logic with pooch in this PR?

### CODE REVIEW

File: tests/test_patch_extraction.py [hunk 1]
```diff
@@ -48,12 +49,10 @@ def read_points_patches(
...
 
+    patches.n = 1870
     with pytest.raises(StopIteration):
-        patches.n = 1870
-        try:
-            next(patches)
-        except StopIteration:
-            raise StopIteration("Index out of bounds.")
+        # skipqc: PTC-W0063
+        next(patches)
```
Comment (vqdang) [inline tests/test_patch_extraction.py:55]: skipqc => skipcq
Alrd fixed in PR128

File: tiatoolbox/data/remote_samples.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
+%YAML 1.2
+---
+aliases:
+  tia: &tia "https://tiatoolbox.dcs.warwick.ac.uk"
+  wsis: &wsis "https://tiatoolbox.dcs.warwick.ac.uk/sample_wsis"
+  wsis2: &wsis2 "https://tiatoolbox.dcs.warwick.ac.uk/testdata/models/new"
```
Comment (vqdang) [inline tiatoolbox/data/remote_samples.yaml:6]: Copied over from chat

```
  wsis: &wsis "https://tiatoolbox.dcs.warwick.ac.uk/sample_wsis"
  modelroot: &modelroot "https://tiatoolbox.dcs.warwick.ac.uk/testdata/models"
```

File: tiatoolbox/data/remote_samples.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  wsi2_4k_4k_svs:
+    url: [*wsis2, "wsi2_4k_4k.svs"]
+  wsi2_4k_4k_jp2:
+    url: [*wsis2, "wsi2_4k_4k.jp2"]
+  wsi2_4k_4k_jpg:
+    url: [*wsis2, "wsi2_4k_4k.jpg"]
+  wsi2_4k_4k_msk:
+    url: [*wsis2, "wsi2_4k_4k.mask.png"]
+  wsi2_4k_4k_pred:
+    url: [*wsis2, "wsi2_4k_4k.pred.dat"]
```
Comment (vqdang) [inline tiatoolbox/data/remote_samples.yaml:75]:
```
  sample-patch-1:
    url: [*modelroot, "samples/kather_patch1.tif"]
  sample-patch-2:
    url: [*modelroot, "samples/kather_patch2.tif"]
  wsi1_8k_8k_svs:
    url: [*modelroot, "samples/wsi1_8k_8k.svs"]
  wsi1_8k_8k_jp2:
    url: [*modelroot, "samples/wsi1_8k_8k.jp2"]
  wsi1_8k_8k_jpg:
    url: [*modelroot, "samples/wsi1_8k_8k.jpg"]
  wsi2_4k_4k_svs:
    url: [*modelroot, "samples/wsi2_4k_4k.svs"]
  wsi2_4k_4k_jp2:
    url: [*modelroot, "samples/wsi2_4k_4k.jp2"]
  wsi2_4k_4k_jpg:
    url: [*modelroot, "samples/wsi2_4k_4k.jpg"]
  wsi2_4k_4k_msk:
    url: [*modelroot, "samples/wsi2_4k_4k.mask.png"]
  wsi2_4k_4k_pred:
    url: [*modelroot, "samples/wsi2_4k_4k.pred.dat"]
  wsi3_20k_20k_svs:
    url: [*modelroot, "samples/wsi3_20k_20k.svs"]
  wsi4_4k_4k_svs:
    url: [*modelroot, "samples/wsi3_20k_20k.svs"]
  wsi3_20k_20k_pred:
    url: [*modelroot, "predictions/tissue_mask/wsi3_20k_20k.mask.png"]
  wsi4_4k_4k_pred:
    url: [*modelroot, "predictions/bcss/wsi4_4k_4k.mask.npy"]
  ```

**Review by vqdang — CHANGES_REQUESTED** [review]

File: tests/test_patch_extraction.py [hunk 4]
```diff
@@ -549,8 +546,10 @@ def test_mask_based_patch_extractor_ndpi(_sample_ndpi):
     )
 
     # Test passing an empty mask
-    with pytest.raises(ValueError):
-        wsi_mask = np.zeros(mask_dim, dtype=np.uint8)
+    wsi_mask = np.zeros(mask_dim, dtype=np.uint8)
+    with pytest.raises(
+        ValueError, match="`mask_reader` should be wsireader.VirtualWSIReader."
+    ):
         patches = patchextraction.get_patch_extractor(
```
Comment (vqdang) [inline tests/test_patch_extraction.py:553]: Fix:
```
    with pytest.raises(
        ValueError, match=r".*No candidate coordinates left.*"
    ):
```

File: tests/test_utils.py [hunk 5]
```diff
@@ -964,7 +970,7 @@ def test_crop_and_pad_edges_non_positive_bounds_size():
 
 def test_normalise_padding_input_dims():
     """Test that normalise padding error with input dimensions > 1."""
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="1 dimensional"):
         utils.image.normalise_padding_size(((0, 0), (0, 0)))
```
Comment (vqdang) [inline tests/test_utils.py:975]: Will crash, need to change `"1 dimensional"` to dimensional/dimentional either in `normalise_padding_size` or in pytest.

File: tests/test_utils.py [hunk 6]
```diff
@@ -952,7 +958,7 @@ def test_crop_and_pad_edges_non_positive_bounds_size():
             pad_mode="constant",
         )
 
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="ounds.*must be >= 0"):
         # Zero dimensions and negative bounds size
```
Comment (vqdang) [inline tests/test_utils.py:962]: >with pytest.raises(ValueError, match="[bB]ounds.*> 0"):

**Review by vqdang — CHANGES_REQUESTED** [review]
Minor fixed required, also travis didnt get triggered.

File: tiatoolbox/data/__init__.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import pathlib
```
Comment (shaneahmed) [inline tiatoolbox/data/__init__.py:1]: do we need the license blurb at the top?

Comment (shaneahmed) [reply tiatoolbox/data/__init__.py:1]: and some docstring for this file?

File: tiatoolbox/data/remote_samples.yaml (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/data/remote_samples.yaml:75]: I think modelroot should be to fetch models. wsis should be in wsi.

Comment (John-P) [reply tiatoolbox/data/remote_samples.yaml]: Have these files been moved on the server?

**Pull request merged**